### PR TITLE
MD-661, ZENKO-945: fix delimiterMaster::filter

### PR DESCRIPTION
**Pull request by jeremyds**
The return values of the delimiterMaster::filter function are used by
its client, metadata back-end like the mongoClient or Metadata, to
implement a skipScan mechanism: after a number of consecutive SKIP
return values, the clients changes their dataset to the next key range.

This algo, by allowing to skip values unwanted in the results
efficiently gives good performance to master version listing.

The previous algo was broken, preventing it client to perform the
skipScan: it returns ACCEPT for values like PHD, delete marker, and
versions, reseting the SKIP counter of clients.

This commit changes the return values of this function to allow
delimiter clients to use the skipScan mechanism.

### CODE REVIEW

**Comment by jeremyds** [comment]
@bert-e wait

File: lib/algos/list/delimiterMaster.js
Comment (rahulreddy) [inline lib/algos/list/delimiterMaster.js]: Isn't this conditional `key === this.NextMarker` already taken care of on line 43?

File: lib/algos/list/delimiterMaster.js
Comment (rahulreddy) [inline lib/algos/list/delimiterMaster.js]: ~~Doesn't this mean PHD along with master gets added to the result set?~~

File: tests/unit/algos/list/delimiterMaster.js
Comment (rahulreddy) [inline tests/unit/algos/list/delimiterMaster.js]: This can be split into two different tests one testing with NextMarker and the other one without it.

File: tests/unit/algos/list/delimiterMaster.js
Comment (rahulreddy) [inline tests/unit/algos/list/delimiterMaster.js]: This is the normal case. It can be written as a test itself. This test should be testing if it's skipping PHD.

File: tests/unit/algos/list/delimiterMaster.js
Comment (rahulreddy) [inline tests/unit/algos/list/delimiterMaster.js]: This can be `should skip a delete marker`

File: tests/unit/algos/list/delimiterMaster.js
Comment (rahulreddy) [inline tests/unit/algos/list/delimiterMaster.js]: `should accept a version`

File: tests/unit/algos/list/delimiterMaster.js
Comment (rahulreddy) [inline tests/unit/algos/list/delimiterMaster.js]: You can assert.deepStrictEqual instead of converting them into strings.

**Comment by jeremyds** [comment]
this Arsenal branch tested in : 
- S3: [here](https://github.com/scality/cloudserver/commits/wip/ZENKO-945-delimitermaster-filter)
- MD:[here]( https://github.com/scality/MetaData/tree/wip/ZENKO-945-delimitermaster-filter)

File: tests/unit/algos/list/delimiterMaster.js
Comment (rahulreddy) [inline tests/unit/algos/list/delimiterMaster.js]: Can you add an assertion showing that it has been added to `CommonPrefixes`?

File: tests/unit/algos/list/delimiterMaster.js
Comment (rahulreddy) [inline tests/unit/algos/list/delimiterMaster.js]: This can be assert.deepStrictEqual instead of stringifying the results.

**Review by rahulreddy — CHANGES_REQUESTED** [review]
In the tests whenever `.filter` is called, it changes the state of the delimiter instance. I recommend confining the tests to be specific, i.e. a unit test should test only one thing.

Can you add a test case where the entries may come out of order, for example
```
delimiter.filter(versionedKey);
delimiter.filer(deletemarker);
delimiter.filter(masterKey);
delimiter.filter(PHDKey);
```

File: lib/algos/list/delimiterMaster.js
Comment (jeremyds) [reply lib/algos/list/delimiterMaster.js]: not exactly, here key is the object name without the version suffix. It handles a specific case and ~~may~~ deserve a comment: if the previous listing page ends with the major version, the next page will have the nextMarker set to the master key. In that case, prvPHDKey is not set (undefined default value) and comparing the nextMarker is the only way to identify this case.

File: tests/unit/algos/list/delimiterMaster.js
Comment (jeremyds) [reply tests/unit/algos/list/delimiterMaster.js]: you re right, it will ease the read

File: tests/unit/algos/list/delimiterMaster.js
Comment (jeremyds) [reply tests/unit/algos/list/delimiterMaster.js]: done

File: lib/algos/list/delimiterMaster.js
Comment (jeremyds) [reply lib/algos/list/delimiterMaster.js]: comment added

**Review by MathieuCassagne — APPROVED** [review]

File: lib/algos/list/delimiterMaster.js
Comment (lamphamsy) [inline lib/algos/list/delimiterMaster.js]: I think it's better to rename `prvPHDKey` according to the change :)

**Review by lamphamsy — APPROVED** [review]
LGTM

File: lib/algos/list/delimiterMaster.js
Comment (lamphamsy) [inline lib/algos/list/delimiterMaster.js]: I found sth after my approval.
This point and https://github.com/scality/Arsenal/pull/550/files#diff-d72bb66c5574d68c654584b499fe9b9aR80 return `FILTER_ACCEPT`, then will reset the skip-scan mechanism. Hence, as the skipping of the same common prefix is performed later: https://github.com/scality/Arsenal/pull/550/files#diff-d72bb66c5574d68c654584b499fe9b9aR95, we will list (all) objects that are either `isPHD` (Master) or `isDeleteMarker` (Version) for a comment prefix.

Comment (lamphamsy) [reply lib/algos/list/delimiterMaster.js]: To avoid this issue, could we move the block that check `commonPrefixes` (https://github.com/scality/Arsenal/pull/550/files#diff-d72bb66c5574d68c654584b499fe9b9aR89) to be performed before checking `isPHD` (https://github.com/scality/Arsenal/pull/550/files#diff-d72bb66c5574d68c654584b499fe9b9aR72)?

Comment (jeremyds) [reply lib/algos/list/delimiterMaster.js]: Returning FILTER_ACCEPT to reset the skip mechanism here is intentional, we want to avoid the client to skip the next entries. 
We reach these two tests when the first one (line 55 & 67) fails to skip the entry, meaning either the entry key does not have a version part or the key has a version and the key part is different from the previous master accepted.
- If the key does have a version part, it is the master version and it may be a PHD. When it is the case we don't want to call addContents or addCOmmonPrefix with this value and, to be able to find the latest version of this key, we want to avoid the client skipping the next entries in its current key range.
- If the entry key has a version part but the entry has not been skipped with the previous test, it is the same principle, we don't want to add it to result and we want to be sure to have a find the latest version.

This kind of input leads to a buggy result by returning FILTER_SKIP in these cases:

> given the SKIP number to activate the skip scan mechanism is SKIP_NUMBER
> a listing page ending with SKIP_NUMBER -1 FILTER_SKIP results
> the next listing page starting with an entry either a PHD or a delete marker

Here, if we return SKIP for the first entry of the second listing, it will trigger the skip scan mechanism and we will miss the version for this key.

Or I miss something :(
At least I should comment more the code, some corner cases are still non obvious.

File: lib/algos/list/delimiterMaster.js
Comment (jeremyds) [reply lib/algos/list/delimiterMaster.js]: right!

File: lib/algos/list/delimiterMaster.js
Comment (rahulreddy) [inline lib/algos/list/delimiterMaster.js]: ~~metata~~ metadata

Comment (jeremyds) [reply lib/algos/list/delimiterMaster.js]: oops

Comment (jeremyds) [reply lib/algos/list/delimiterMaster.js]: done

File: tests/unit/algos/list/delimiterMaster.js
Comment (rahulreddy) [inline tests/unit/algos/list/delimiterMaster.js]: ~~thedelimiter~~ the delimiter

Comment (jeremyds) [reply tests/unit/algos/list/delimiterMaster.js]: done

**Review by vrancurel — APPROVED** [review]

**Review by rahulreddy — APPROVED** [review]

File: lib/algos/list/delimiterMaster.js
Comment (MathieuCassagne) [reply lib/algos/list/delimiterMaster.js]: We investigated with @lamphamsy, and we found out that the extension returns the expected result, but resetting the skip scan when a PHD/isDeleteMarker version is encountered can make it not working at all, depending on the data inside the bucket.

For example you have 10000 master key in a bucket, but with every 90 master key is a PHD one.
You want to do a listBucket operation, the extension will return FILTER_ACCEPT for each PHD version it encounters, which reset the skip scan. So every 90 master keys the skip scan get resetted, and we end up listing the 10000keys instead of ~100.

I think a good solution would be to have a FILTER_CONTINUE flag, that will go to the next entry without resetting the streakLength counter, and use it when we encounter a PHD/isDeleteMarker version.

Comment (lamphamsy) [reply lib/algos/list/delimiterMaster.js]: I add sth to @MathieuCassagne 's example: All the keys have a same prefix, e.g. `prefix1/`. The expected result of a list operation should contain an array of common prefixes that is exactly [`prefix1`]. Thanks to the skip-scan mechanism, we should obtain the expected result after scanning several hundreds (200 in Metadata) of keys.
In the case @MathieuCassagne mentioned above, we have to scan all keys as skip-scan's counter is reset for every a PHD Master key.

**Comment by lamphamsy** [comment]
There is one more case we should handle:

Given an object having several versions, e.g. `v1`, `v2`, we delete it without specifying versionId => we delete its Master key (if I understand correctly) => 
- Set Master key` isDeleteMarker = true` (but not `isPHD`),  and
- All its version keys `v1`, `v2` are not changed.

When we list the bucket, keys of the object are received:
- Master key `isDeleteMarker = true` and `isPHD = false` => return `FILTER_ACCEPT`. And `prvKey` is still `underfined` or equal to Master key of the previously listed object: https://github.com/scality/Arsenal/blob/464ab87903bf17962774ef523b1ea374ea43ad48/lib/algos/list/delimiterMaster.js#L84
- Version key `v1` => pass all checks, go to https://github.com/scality/Arsenal/blob/464ab87903bf17962774ef523b1ea374ea43ad48/lib/algos/list/delimiterMaster.js#L87 => add it to the result => It should not be there.

File: lib/algos/list/delimiterMaster.js
Comment (jeremyds) [reply lib/algos/list/delimiterMaster.js]: You re absolutely right, we missed this use case. It should be fixed.

Comment (jeremyds) [reply lib/algos/list/delimiterMaster.js]: my previous comment was addressed to your other comment. about the incorrect listing result.

**Review by rahulreddy — COMMENTED** [review]
Seems to cover the issues raised. @MathieuCassagne and @lamphamsy . Can you review and approve if it works. You can remove berte wait comment to merge.

File: lib/algos/list/delimiterMaster.js
Comment (lamphamsy) [inline lib/algos/list/delimiterMaster.js]: As `this.NextMarker` can be either
- case1: its Master key if that key is added by `addContents()`: https://github.com/scality/Arsenal/blob/2db904952bddcc9a40f786444db44514ae29908a/lib/algos/list/delimiter.js#L132 
- case2: only a common prefix extracted from a Master key added by addCommonPrefix(): https://github.com/scality/Arsenal/blob/2db904952bddcc9a40f786444db44514ae29908a/lib/algos/list/delimiter.js#L184

Case1 is checked correctly here but Case2 might be not. Should we check this case?

File: lib/algos/list/delimiterMaster.js
Comment (lamphamsy) [inline lib/algos/list/delimiterMaster.js:91]: 👍

**Review by lamphamsy — COMMENTED** [review]
Do we skip the case mentioned in https://github.com/scality/Arsenal/pull/550#discussion_r210588132?
Moreover, there is a small case that might be verified.

File: lib/algos/list/delimiterMaster.js
Comment (MathieuCassagne) [reply lib/algos/list/delimiterMaster.js]: For the case 2, if a master key has been added by addCommonPrefix(), this.prvKey is set to the master key, and so the check at line 67 is true, and we return FILTER_SKIP.
I think we're good ?

Comment (lamphamsy) [reply lib/algos/list/delimiterMaster.js]: yes, it's true but only if these keys are on the same page.
We need `this.NextMarker` only for the next page starts: https://github.com/scality/Arsenal/blob/2db904952bddcc9a40f786444db44514ae29908a/lib/algos/list/delimiterMaster.js#L60
At that time, `this.prvKey` is reset to be `undefined` in the constructor.

Comment (lamphamsy) [reply lib/algos/list/delimiterMaster.js]: I think we could add an extra condition that checks `NextMarker` is a common prefix and `key` starting with `NextMarker`, e.g.

```javascript
if (key === this.prvKey || key === this.NextMarker ||
    (this.delimiter && this.NextMarker && 
     this.NextMarker.endsWith(this.delimiter) && key.startsWith(this.NextMarker)) {
```

Comment (MathieuCassagne) [reply lib/algos/list/delimiterMaster.js]: After more investigation it appear that our NextMarker might not be set to the good value when it is a CommonPrefixes, we're setting it to the common prefix we found, but we should set it to the key which we extract the common prefix.

The AWS documentation about NextMarker:
"
When the response is truncated (that is, the IsTruncated element value in the response is true), you can use the key name in this field as a marker in the subsequent request to get next set of objects. Amazon S3 lists objects in UTF-8 character encoding in lexicographical order. 
"
It is said that NextMarker, if exists, should be a key, but we set it to a CommonPrefix, which is not a key.
If we can change this, the check line 67, is good as it is for both case 1 and 2.

Also there is a transformation done on NextMarker in the constructor of delimiter.js, we might want to use the marker(which is NextMarker we sent back without the modification) parameter instead.

Comment (jeremyds) [reply lib/algos/list/delimiterMaster.js]: wow, amazing review again. I don't want to handle the change of NextMarker value for common prefix in delimiter.js in this PR. I am going to open an issue, adds the test mentioned by @lamphamsy here https://github.com/scality/Arsenal/pull/550#discussion_r211982038 and tracked this issue by a comment in the code.

**Review by rahulreddy — APPROVED** [review]
I cannot review with the same depth as @MathieuCassagne and @MathieuCassagne (thanks guys for the reviews and raising the issues). Going by their reviews, doing my own review and seeing that @jeremyds addressed the corner case raised, I am merging this PR. 
@jeremyds will follow up with the common prefix fix for delimiter.js in another PR.

**Pull request merged**
